File: combined_approach/robot_interface_node.py
import rclpy
from rcl_interfaces.srv import GetParameters
from rclpy.node import Node
from robot_model import Model, create_urdf_from_string
from sensor_msgs.msg import JointState as JointStateMsg
from state_representation import JointState
from std_msgs.msg import Float64MultiArray


class RobotInterfaceNode(Node):
    def __init__(self, node_name, subscription_topic, publisher_topic):
        super().__init__(node_name)
        self.declare_parameter("robot_name", "robot")
        self.publisher = self.create_publisher(Float64MultiArray, publisher_topic, 10)
        self.robot = None
        self.joint_state = None
        self.state_received = False

        self._subscription_topic = subscription_topic
        self._subscription = None

    def init_robot_model(self, name):
        client = self.create_client(
            GetParameters, "/" + name + "/robot_state_publisher/get_parameters"
        )
        while not client.wait_for_service(1):
            if not rclpy.ok():
                self.get_logger().error(
                    "Interrupted while waiting for the service. Exiting."
                )
                rclpy.shutdown()
            self.get_logger().info("Service not available, waiting again...")
        request = GetParameters.Request()
        request.names = ["robot_description"]
        future = client.call_async(request)
        rclpy.spin_until_future_complete(self, future)
        robot_description = future.result().values[0].string_value
        urdf_path = "/tmp/" + name + ".urdf"
        create_urdf_from_string(robot_description, urdf_path)
        self.robot = Model(name, urdf_path)
        self.joint_state = JointState(
            self.robot.get_robot_name(), self.robot.get_joint_frames()
        )
        self._subscription = self.create_subscription(
            JointStateMsg, self._subscription_topic, self._robot_state_callback, 10
        )

    def _robot_state_callback(self, msg):
        if not self.state_received:
            self.state_received = True
        self.joint_state.set_positions(msg.position)
        self.joint_state.set_velocities(msg.velocity)
        self.joint_state.set_torques(msg.effort)

File: src/RobotInterfaceNode.cpp
#include "RobotInterfaceNode.h"

namespace ros2_examples {

RobotInterfaceNode::RobotInterfaceNode(
    const std::string& node_name, const std::string& subscription_topic, const std::string& publisher_topic
) : Node(node_name, rclcpp::NodeOptions().use_intra_process_comms(false)), subscription_topic_(subscription_topic) {
  this->declare_parameter<std::string>("robot_name", "robot");
  this->publisher = this->create_publisher<std_msgs::msg::Float64MultiArray>(publisher_topic, 10);
}

void RobotInterfaceNode::robot_state_callback(const sensor_msgs::msg::JointState::SharedPtr msg) {
  if (!this->state_received) {
    this->state_received = true;
  }
  this->joint_state.set_positions(msg->position);
  this->joint_state.set_velocities(msg->velocity);
  this->joint_state.set_torques(msg->effort);
}

void RobotInterfaceNode::init_robot_model(const std::string& name) {
  auto parameters_client = std::make_shared<rclcpp::SyncParametersClient>(this, "robot_state_publisher");
  while (!parameters_client->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
      rclcpp::shutdown();
    }
    RCLCPP_INFO(this->get_logger(), "Service not available, waiting again...");
  }
  auto urdf_string = parameters_client->get_parameter<std::string>("robot_description");

  std::string urdf_path = "/tmp/" + name + ".urdf";
  robot_model::Model::create_urdf_from_string(urdf_string, urdf_path);
  this->robot = std::make_shared<robot_model::Model>(name, urdf_path);
  this->joint_state = state_representation::JointState(this->robot->get_robot_name(), this->robot->get_joint_frames());
  this->subscription_ = this->create_subscription<sensor_msgs::msg::JointState>(
      this->subscription_topic_, 10, std::bind(&RobotInterfaceNode::robot_state_callback, this, std::placeholders::_1));
}
}// namespace ros2_examples
